### PR TITLE
Consistify exiting draw modes

### LINKED ISSUE #399: Finishing a feature behaves unexpectedly
## Desired functionality 

- **Escape** should always destroy the feature
- **Enter** should clear the feature if at the last click the feature was not valid, other wise it should remove the currently being added leg and save the feature.
- **1-3** should behave like `enter`

## Current functionality

- Before one click
  - All features behave the same.
  -  clicking `escape`, `enter` or `1-3` results in no feature being added to the map
- After first click (LineString, Polygon)
  - LineString
    - `escape` removes the feature
    - `enter` removes the feature
    - `1-3` keeps the feature :bug:
  - Polygon
    - clicking `escape`, `enter` or `1-3` results in no feature being added to the map
- After second click
  - LineString
    - `escape` removes the feature
    - `enter` keeps one leg
    - `1-3` keeps the two legs :bug:
  - Polygon
    - `escape` removes the feature
    - `enter` removes the feature and throws an error :bug:
    - `1-3` keeps the feature :bug:
- After 3 or more clicks
  - LineString
    - `escape` removes the feature
    - `enter` removes most recently dragged out leg
    - `1-3` keeps all legs :bug:
  - Polygon
    - `escape` removes the feature
    - `enter` removes most recently dragged out leg
    - `1-3` keeps all legs :bug:

/cc @davidtheclark @samanpwbb

### PR DESCRIPTION
**Pull request by mcwhittemore**
Closes #399

@davidtheclark for the review

### CODE REVIEW

File: src/modes/draw_line_string.js [hunk 1]
```diff
@@ -60,17 +45,28 @@ module.exports = function(ctx) {
...
+      });
     },
 
     stop() {
       doubleClickZoom.enable(ctx);
       ctx.ui.setActiveButton();
 
-      // If it's invalid, just destroy the thing
-      if (!line.isValid()) {
+      line.removeCoordinate(`${currentVertexPosition}`);
```
Comment (davidtheclark) [inline src/modes/draw_line_string.js]: Are we able to remove the coordinate if the line is not valid?

File: src/modes/draw_line_string.js [hunk 2]
```diff
@@ -60,17 +45,28 @@ module.exports = function(ctx) {
...
+      if (line.isValid() && ctx.store.get(line.id) !== undefined) {
```
Comment (davidtheclark) [inline src/modes/draw_line_string.js]: When can you arrive here with an undefined feature?

File: src/modes/draw_line_string.js [hunk 3]
```diff
@@ -60,17 +45,28 @@ module.exports = function(ctx) {
...
+        ctx.map.fire(Constants.events.CREATE, {
+          features: [line.toGeoJSON()]
+        });
+      }
+      else if (ctx.store.get(line.id) !== undefined) {
         ctx.store.delete([line.id], { silent: true });
+        ctx.events.changeMode(Constants.modes.SIMPLE_SELECT, {}, { silent: true });
```
Comment (davidtheclark) [inline src/modes/draw_line_string.js]: I think the repetition of these two lines in 3 places justifies something like the `finish` function, right?

File: src/modes/draw_polygon.js [hunk 4]
```diff
@@ -58,29 +43,41 @@ module.exports = function(ctx) {
...
+      });
     },
 
     stop: function() {
       doubleClickZoom.enable(ctx);
       ctx.ui.setActiveButton();
 
-      // If it's invalid, just destroy the thing
-      if (!polygon.isValid()) {
+      polygon.removeCoordinate(`0.${currentVertexPosition}`);
```
Comment (davidtheclark) [inline src/modes/draw_polygon.js]: I think I ran into a bug where this failed when the polygon was not valid.

File: src/modes/draw_polygon.js [hunk 5]
```diff
@@ -58,29 +43,41 @@ module.exports = function(ctx) {
...
+      if (polygon.isValid() && ctx.store.get(polygon.id) !== undefined) {
+        ctx.map.fire(Constants.events.CREATE, {
+          features: [polygon.toGeoJSON()]
+        });
+      }
+      else if (ctx.store.get(polygon.id) !== undefined) {
         ctx.store.delete([polygon.id], { silent: true });
+        ctx.events.changeMode(Constants.modes.SIMPLE_SELECT, {}, { silent: true });
       }
```
Comment (davidtheclark) [inline src/modes/draw_polygon.js]: These conditionals could be simplified a little by early returning if the feature is undefined:

```js
if (ctx.store.get(polygon.id) === undefied) return;
if (polygon.isValid()) {
 ..
} else {
 ..
}
```

File: test/draw_line_string.test.js [hunk 6]
```diff
@@ -295,52 +276,51 @@ test('draw_line_string interaction', t => {
...
-      click(map, makeMouseEvent(1, 1));
+      mouseClick(map, makeMouseEvent(1, 1));
+      map.fire('mousemove', makeMouseEvent(16, 16));
       st.equal(Draw.getAll().features.length, 1, 'no longer drawing');
 
       st.end();
     });
 
-    t.test('start draw_line_string mode then exit before a click', st => {
+    t.test('start draw_line_string mode then change_mode before a click', st => {
```
Comment (davidtheclark) [inline test/draw_line_string.test.js]: change mode without underscore, right? underscore makes it seem like it's a mode

File: test/utils/key_events.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+export const startPointEvent = createSyntheticEvent('keydown', {
+  keyCode: 49
+});
+
+export const startLineStringEvent = createSyntheticEvent('keydown', {
+  keyCode: 50
+});
+
+export const startPolygonEvent = createSyntheticEvent('keydown', {
+  keyCode: 51
```
Comment (davidtheclark) [inline test/utils/key_events.js]: Are the 1-3 keyboard interactions documented? I don't see them.

File: src/modes/draw_line_string.js (hunk 1 above)
Comment (mcwhittemore) [inline src/modes/draw_line_string.js]: yep. tests cover both cases.

File: src/modes/draw_line_string.js (hunk 2 above)
Comment (mcwhittemore) [inline src/modes/draw_line_string.js]: Line is a local scoped var so deleting the line from the store doesn't kill this. If we hit escape and get here the line could be valid be deleted.

File: src/modes/draw_polygon.js (hunk 4 above)
Comment (mcwhittemore) [inline src/modes/draw_polygon.js]: Can you explain this a bit more. I'm not seeing it.

Comment (davidtheclark) [inline src/modes/draw_polygon.js]: If I remember right, I think I maybe have hit this error when I entered polygon draw mode then double clicked on the map. May have just been an error I saw in the console, not something that broke the app.

File: src/modes/draw_line_string.js (hunk 2 above)
Comment (davidtheclark) [inline src/modes/draw_line_string.js]: Makes sense. A comment in the code would be nice to explain that context.

File: test/utils/key_events.js (hunk 7 above)
Comment (mcwhittemore) [inline test/utils/key_events.js]: Hmmm... these are part of the point, line, and polygon controls. We should fill out those docs.

**Comment by davidtheclark** [comment]
These are great tests and code cleanups! I had some questions that you see.

File: test/utils/key_events.js (hunk 7 above)
Comment (davidtheclark) [inline test/utils/key_events.js]: If you want to just make another issue for that, that'd be swell.

Comment (mcwhittemore) [inline test/utils/key_events.js]: - [x] done

File: src/modes/draw_polygon.js (hunk 4 above)
Comment (mcwhittemore) [inline src/modes/draw_polygon.js]: hmmm... I'm not seeing this locally.
check

Comment (davidtheclark) [inline src/modes/draw_polygon.js]: Ok, maybe something else has taken care of that. If it's not happening, let's forget about it.

**Comment by mcwhittemore** [comment]
@davidtheclark tests failed once because of a flap... looks like maybe something with `draw.render` events not being fired in a test where we expect it to and than never ending. I'm not sure if its something about the changes here, just want to make sure we have our eyes out for it so we can crush it when/if it flaps again.

**Pull request merged**